ext/mbstring/php_unicode.c

```diff
@@ -229,195 +228,144 @@ static inline unsigned php_unicode_tofold_simple(unsigned code, enum mbfl_no_enc
-					code = php_unicode_totitle_simple(c, data->no_encoding);
-					(data->next_filter->filter_function)(code, data->next_filter);
+				w = php_unicode_tolower_raw(w, enc);
+				if (UNEXPECTED(w > 0xFFFFFF)) {
+					p = emit_special_casing_sequence(w, p);
 				} else {
-					php_unicode_totitle_full(c, data->no_encoding, data->next_filter);
+					*p++ = w;
 				}
 			}
-			if (!php_unicode_is_case_ignorable(c)) {
-				data->title_mode = php_unicode_is_cased(c);
-			}
 			break;
-		}
-		EMPTY_SWITCH_DEFAULT_CASE()
-	}
-
-	return 0;
-}
 
-MBSTRING_API char *php_unicode_convert_case(
-		int case_mode, const char *srcstr, size_t srclen, size_t *ret_len,
-		const mbfl_encoding *src_encoding, int illegal_mode, int illegal_substchar)
-{
-	struct convert_case_data data;
-	mbfl_convert_filter *from_wchar, *to_wchar;
-	mbfl_string result;
-
-	mbfl_memory_device device;
-	mbfl_memory_device_init(&device, srclen + 1, 0);
-
-	/* encoding -> wchar filter */
-	to_wchar = mbfl_convert_filter_new(src_encoding,
-			&mbfl_encoding_wchar, convert_case_filter, NULL, &data);
-	if (to_wchar == NULL) {
-		mbfl_memory_device_clear(&device);
-		return NULL;
-	}
-
-	/* wchar -> encoding filter */
-	from_wchar = mbfl_convert_filter_new(
-			&mbfl_encoding_wchar, src_encoding,
-			mbfl_memory_device_output, NULL, &device);
-	if (from_wchar == NULL) {
-		mbfl_convert_filter_delete(to_wchar);
-		mbfl_memory_device_clear(&device);
-		return NULL;
-	}
+		case PHP_UNICODE_CASE_FOLD:
+			for (int i = 0; i < out_len; i++) {
+				uint32_t w = wchar_buf[i];
+				if (UNEXPECTED(w > 0xFFFFFF)) {
+					*p++ = w;
+					continue;
+				}
+				w = php_unicode_tofold_raw(w, enc);
+				if (UNEXPECTED(w > 0xFFFFFF)) {
+					p = emit_special_casing_sequence(w, p);
+				} else {
+					*p++ = w;
+				}
+			}
+			break;
 
-	to_wchar->illegal_mode = illegal_mode;
-	to_wchar->illegal_substchar = illegal_substchar;
-	from_wchar->illegal_mode = illegal_mode;
-	from_wchar->illegal_substchar = illegal_substchar;
-
-	data.next_filter = from_wchar;
-	data.no_encoding = src_encoding->no_encoding;
-	data.case_mode = case_mode;
-	data.title_mode = 0;
-
-	{
-		/* feed data */
-		const unsigned char *p = (const unsigned char *) srcstr;
-		size_t n = srclen;
-		while (n > 0) {
-			if ((*to_wchar->filter_function)(*p++, to_wchar) < 0) {
-				break;
+		case PHP_UNICODE_CASE_TITLE:
+			for (int i = 0; i < out_len; i++) {
+				uint32_t w = wchar_buf[i];
+				if (UNEXPECTED(w > 0xFFFFFF)) {
+					*p++ = w;
+					continue;
+				}
+				uint32_t w2 = title_mode ? php_unicode_tolower_raw(w, enc) : php_unicode_totitle_raw(w, enc);
+				if (UNEXPECTED(w2 > 0xFFFFFF)) {
+					p = emit_special_casing_sequence(w2, p);
+				} else {
+					*p++ = w2;
+				}
+				if (!php_unicode_is_case_ignorable(w)) {
+					title_mode = php_unicode_is_cased(w);
+				}
 			}
-			n--;
+			break;
+
+		default:
+			ZEND_ASSERT(0 && "We should not reach here; all valid values of case_mode were checked above");
```

Comment:
That's what the ``EMPTY_SWITCH_DEFAULT_CASE()`` does, assert that it doesn't land there.

Reply:
Thanks! Adjusted to use `EMPTY_SWITCH_DEFAULT_CASE`.